src/lib/signature.ts

```diff
@@ -208,10 +225,51 @@ class Signature extends CircuitValue {
    */
   verify(publicKey: PublicKey, msg: Field[]): Bool {
     const point = publicKey.toGroup();
-    let e = Scalar.fromBits(
-      Poseidon.hash(msg.concat([point.x, point.y, this.r])).toBits()
+    let h = hashWithPrefix(
+      prefixes.signatureTestnet,
+      msg.concat([point.x, point.y, this.r])
     );
-    let r = point.scale(e).neg().add(Group.generator.scale(this.s));
+    // TODO: Scalar.fromBits interprets the input as a "shifted scalar"
+    // therefore we have to use scaleShifted which is very inefficient
+    let e = Scalar.fromBits(h.toBits());
+    let r = scaleShifted(point, e).neg().add(Group.generator.scale(this.s));
     return Bool.and(r.x.equals(this.r), r.y.toBits()[0].equals(false));
   }
+
+  /**
+   * Decodes a base58 encoded signature into a {@link Signature}.
+   */
+  static fromBase58(signatureBase58: string) {
+    let { r, s } = SignatureBigint.fromBase58(signatureBase58);
+    return Signature.fromObject({
+      r: Field(r),
+      s: Scalar.fromJSON(s.toString()),
+    });
+  }
+  /**
+   * Encodes a {@link Signature} in base58 format.
+   */
+  toBase58() {
+    let r = this.r.toBigInt();
+    let s = BigInt(this.s.toJSON());
+    return SignatureBigint.toBase58({ r, s });
+  }
+}
+
+// performs scalar multiplication s*G assuming that instead of s, we got s' = 2s + 1 + 2^255
+// cost: 2x scale by constant, 1x scale by variable
+function scaleShifted(point: Group, shiftedScalar: Scalar) {
+  let oneHalfGroup = point.scale(Scalar.fromBigInt(oneHalf));
+  let shiftGroup = oneHalfGroup.scale(Scalar.fromBigInt(shift));
+  return oneHalfGroup.scale(shiftedScalar).sub(shiftGroup);
+}
+// returns s, assuming that instead of s, we got s' = 2s + 1 + 2^255
+// (only works out of snark)
+function unshift(shiftedScalar: Scalar) {
+  return shiftedScalar
+    .sub(Scalar.fromBigInt(shift))
+    .mul(Scalar.fromBigInt(oneHalf));
 }
```

Comment:
why does this only work out of snark? Also, isn't this a problem if I want to sign within a snark? (since it's used in the signature algo)

Reply:
only works out of snark because `sub` and `mul` are not available on scalars. not sure how efficient they would be to implement, compared to the extra scaling I have to do instead. do you know?

the signing algorithm was already (I think, deliberately) written to not work inside a snark (for example, it used a constant random field element for the nonce 😱 ; now I'm deriving the nonce as a blake2b hash. tbf, the nonce could be a prover witness). the above will throw an error in the snark. 

Inherently, signing in the snark is not as useful because then you have to provide your private key as input -- but that usually sits in some guarded place like your wallet. so, I think it's fine to only support signing outside + verifying inside which achieves the same and plays well with wallets